### DISCUSSION TITLE
GitHub Actions Failing with [Errno 28] No space left on device

### DISCUSSION THREAD

**Discussion by matthewfeickert** [community/community#26488] [Actions]
<p>I have a scheduled GitHub Actions that runs CI nightly for my team’s project that was working fine but is now failing with</p>
<p><a href="https://github.com/scikit-hep/pyhf/runs/596983555?check_suite_focus=true#step:4:20" rel="nofollow noopener">ERROR: Could not install packages due to an EnvironmentError: [Errno 28] No space left on device</a></p>
<p>The total amount of installed software in the CI job is less that 3.6 GB so I am not sure why things are suddenly failing. Has there been a silent change in the amount of storage that is available in the GitHub Actions runners?</p>
<p>Here are examples of scheduled CI jobs 1 day apart, one which ran fine and the other errored as above:</p>
<ul>
<li>
<p><a href="https://github.com/scikit-hep/pyhf/actions/runs/80294195" rel="nofollow noopener">Working</a> (2020-04-15)</p>
</li>
<li>
<p><a href="https://github.com/scikit-hep/pyhf/actions/runs/81189933" rel="nofollow noopener">Failing</a> (2020-04-16)</p>
</li>
</ul>

**Comment by viceice** [comment]
<p>Seeing this too <a href="https://github.com/renovatebot/renovate/pull/5996/checks?check_run_id=597383462" rel="nofollow noopener">https://github.com/renovatebot/renovate/pull/5996/checks?check_run_id=597383462</a></p>
<p>ref: <a href="https://github.community/t/diminishing-disk-space-on-ubuntu-images/18150">https://github.community/t5/GitHub-Actions/Diminishing-disk-space-on-Ubuntu-images/m-p/54332#M9137</a></p>

**Comment by rarkins** [comment]
<p>We’ve also experienced problems since yesterday that seem related to disk space on Actions.</p>

**Comment by matthewfeickert** [comment]
<p>Here’s an example where individually installing all the dependencies, that would be installed in the stage that is failing, succeeds (<a href="https://github.com/scikit-hep/pyhf/runs/598685755" rel="nofollow noopener">test (ubuntu-latest, 3.7)</a>) but when they are all installed normally the error occurs (<a href="https://github.com/scikit-hep/pyhf/runs/598685766" rel="nofollow noopener">notebooks (3.7)</a>). We can see (with df -h) from the stage that succeeds that there is still plenty of space available, so this is all very strange.</p>
<p>Before installs:</p>
<pre><code>Filesystem Size Used Avail Use% Mounted on
udev 3.4G 0 3.4G 0% /dev
tmpfs 695M 960K 694M 1% /run
/dev/sda1 84G 75G 8.3G 91% /
tmpfs 3.4G 8.0K 3.4G 1% /dev/shm
tmpfs 5.0M 0 5.0M 0% /run/lock
tmpfs 3.4G 0 3.4G 0% /sys/fs/cgroup
/dev/loop0 40M 40M 0 100% /snap/hub/43
/dev/loop1 94M 94M 0 100% /snap/core/8935
/dev/sda15 105M 3.6M 101M 4% /boot/efi
/dev/sdb1 14G 41M 13G 1% /mnt
</code></pre>
<p>After individual installs:</p>
<pre><code>Filesystem Size Used Avail Use% Mounted on
udev 3.4G 0 3.4G 0% /dev
tmpfs 695M 960K 694M 1% /run
/dev/sda1 84G 79G 5.0G 95% /
tmpfs 3.4G 8.0K 3.4G 1% /dev/shm
tmpfs 5.0M 0 5.0M 0% /run/lock
tmpfs 3.4G 0 3.4G 0% /sys/fs/cgroup
/dev/loop0 40M 40M 0 100% /snap/hub/43
/dev/loop1 94M 94M 0 100% /snap/core/8935
/dev/sda15 105M 3.6M 101M 4% /boot/efi
/dev/sdb1 14G 41M 13G 1% /mnt
</code></pre>

**Comment by matthewfeickert** [comment]
<p>My earlier message got deleted by someone, so reposting this:</p>
<p>We can see an instance where installing all the pacakges individually works (<a href="https://github.com/scikit-hep/pyhf/runs/598685755" rel="nofollow noopener">test (ubunut-latest, 3.7)</a>), but then installing all the packages normally fails (<a href="https://github.com/scikit-hep/pyhf/runs/598685766" rel="nofollow noopener">notebooks (3.7)</a>). We can see (with df -h) in the case that suceeds that there is plenty of space remaining, which makes this all very strange.</p>
<p>Before installs:</p>
<pre><code>Filesystem Size Used Avail Use% Mounted on
udev 3.4G 0 3.4G 0% /dev
tmpfs 695M 960K 694M 1% /run
/dev/sda1 84G 75G 8.3G 91% /
tmpfs 3.4G 8.0K 3.4G 1% /dev/shm
tmpfs 5.0M 0 5.0M 0% /run/lock
tmpfs 3.4G 0 3.4G 0% /sys/fs/cgroup
/dev/loop0 40M 40M 0 100% /snap/hub/43
/dev/loop1 94M 94M 0 100% /snap/core/8935
/dev/sda15 105M 3.6M 101M 4% /boot/efi
/dev/sdb1 14G 41M 13G 1% /mnt
</code></pre>
<p>After individually installing packages:</p>
<pre><code>Filesystem Size Used Avail Use% Mounted on
udev 3.4G 0 3.4G 0% /dev
tmpfs 695M 960K 694M 1% /run
/dev/sda1 84G 79G 5.0G 95% /
tmpfs 3.4G 8.0K 3.4G 1% /dev/shm
tmpfs 5.0M 0 5.0M 0% /run/lock
tmpfs 3.4G 0 3.4G 0% /sys/fs/cgroup
/dev/loop0 40M 40M 0 100% /snap/hub/43
/dev/loop1 94M 94M 0 100% /snap/core/8935
/dev/sda15 105M 3.6M 101M 4% /boot/efi
/dev/sdb1 14G 41M 13G 1% /mnt
</code></pre>

**Comment by matthewfeickert** [comment]
<p><span class="mention">@rarkinscan</span> you also link to the specific run in GitHub Actions that you’re seeing this error? I’ve submitted a help ticket to GitHub Actions and in the event that it gets reviewed by the engineering team it would be nice to have as many examples of this as possible to review.</p>

**Comment by matthewfeickert** [comment]
<p>I noticed something even more strange: this error is only occuring on Linux runners</p>
<pre><code>runs-on: ubuntu-latestjens-maus
</code></pre>
<p>while the macOS builds succeed. I’ve applied <a class="mention" href="/jens-maus">@jens-maus</a>’s hack that fixes things (<a href="https://github.community/t/bug-strange-no-space-left-on-device-ioexceptions-on-github-runners/17616/10">detailed here</a>). I’ve tried this with both removal of /swapfile like they do and found df -h gives</p>
<pre><code>Filesystem Size Used Avail Use% Mounted on
udev 3.4G 0 3.4G 0% /dev
tmpfs 695M 960K 694M 1% /run
/dev/sda1 84G 64G 20G 77% /
tmpfs 3.4G 8.0K 3.4G 1% /dev/shm
tmpfs 5.0M 0 5.0M 0% /run/lock
tmpfs 3.4G 0 3.4G 0% /sys/fs/cgroup
/dev/loop1 40M 40M 0 100% /snap/hub/43
/dev/loop0 94M 94M 0 100% /snap/core/8935
/dev/sda15 105M 3.6M 101M 4% /boot/efi
/dev/sdb1 14G 41M 13G 1% /mnt
</code></pre>
<p>and with just running apt-get clean</p>
<pre><code>Filesystem Size Used Avail Use% Mounted on
udev 3.4G 0 3.4G 0% /dev
tmpfs 695M 960K 694M 1% /run
/dev/sda1 84G 72G 12G 87% /
tmpfs 3.4G 8.0K 3.4G 1% /dev/shm
tmpfs 5.0M 0 5.0M 0% /run/lock
tmpfs 3.4G 0 3.4G 0% /sys/fs/cgroup
/dev/loop0 40M 40M 0 100% /snap/hub/43
/dev/loop1 94M 94M 0 100% /snap/core/8935
/dev/sda15 105M 3.6M 101M 4% /boot/efi
/dev/sdb1 14G 41M 13G 1% /mnt
</code></pre>
<p>For our project it seems that just running with apt-get clean is enough space restored that our Linux runner jobs can proceed again.</p>
<p>However, I have to agree with <a class="mention" href="/saudet">@saudet</a> on the point they raised on “<a href="https://github.community/t/diminishing-disk-space-on-ubuntu-images/18150">Diminishing disk space on Ubuntu images</a>” that this really seems to be a problem. This is targeting only the Linux runners and this started happening silently after months of no issues. Having users go in and manually clean up runners to ensure that there is sufficient space doesn’t really seem like a good solution.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/matthewfeickert">@matthewfeickert</a> ,</p>
<p>Thanks for your sharing! I forked your repo, the disk issue cannot be reproduced on my side now, looks engineering team has fixed it, could you please help to confirm? </p>

**Comment by matthewfeickert** [comment]
<p>Thanks for your reply, <a class="mention" href="/weide-zhou">@weide-zhou</a>. The problem is not fixed though.</p>
<p>You’re not seeing the issue as the fork of pyhf you made includes the hack fix that I mentioned in my previous reply (c.f. <a href="https://github.com/scikit-hep/pyhf/pull/819" rel="nofollow noopener">pyhf PR 819</a>). If you remove the sudo apt-get clean lines the exact same error occurs (<a href="https://github.com/scikit-hep/pyhf/runs/601047906?check_suite_focus=true" rel="nofollow noopener">example of doing exactly that here</a>). For pyhf this can be easily replicated by committing and pushing the results of the following</p>
<pre><code>sed -i '/sudo apt-get clean/d' .github/workflows/ci.yml
</code></pre>

**Comment by konradpabjan** [comment]
<p>For anyone running into issues with no space left, there is an issue open in actions/virtual-environments that has some discussions going on: <a href="https://github.com/actions/virtual-environments/issues/709" rel="nofollow noopener">https://github.com/actions/virtual-environments/issues/709</a></p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/matthewfeickert">@matthewfeickert</a> ,</p>
<p>Thanks for your check! I notice the disscussion in <a href="https://github.com/actions/virtual-environments/issues/709" rel="nofollow noopener">ticket</a> , engineering team did a rollback to the previous version of the VM images used for the virtual-environments in both Ubuntu 16.04 and Ubuntu 18.04.  Could you please confirm whether it works for you?</p>

**Comment by matthewfeickert** [comment]
<p>Yes, the VM rollback resolved the issue. I’ve marked this last post as the “Solution” as you’ve linked to the relevant <a href="https://github.com/actions/virtual-environments/issues/709" rel="nofollow noopener">actions/virtual-environments Issue</a> in it. Thanks!</p>
